include/IECore/Canceller.h

```diff
@@ -0,0 +1,107 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2018, Image Engine Design Inc. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of Image Engine Design nor the names of any
+//       other contributors to this software may be used to endorse or
+//       promote products derived from this software without specific prior
+//       written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef IECORE_CANCELLER_H
+#define IECORE_CANCELLER_H
+
+#include "IECore/Export.h"
+
+#include "boost/noncopyable.hpp"
+
+#include <atomic>
+
+namespace IECore
+{
+
+/// Exception thrown by `Canceller::check()`.
+/// Deliberately _not_ derived from `std::exception`
+/// to minimise the chances of it being accidentally
+/// suppressed or mistaken for an error. In typical
+/// use there is no need to catch exceptions of this
+/// type.
+struct Cancelled
+{
+};
+
+/// Class used to cancel long-running background
+/// operations.
+///
+/// Example :
+///
+/// ```
+/// Canceller c;
+/// thread t(
+///     [&c] {
+///         while( 1 ) {
+///             Canceller::check( &c );
+///         }
+///     }
+/// );
+/// c.cancel();
+/// t.join();
+/// ```
+class IECORE_API Canceller : public boost::noncopyable
+{
+
+	public :
+
+		Canceller()
+			:	m_cancelled( false )
+		{
+		}
+
+		void cancel()
+		{
+			m_cancelled = true;
+		}
+
+		static void check( const Canceller *canceller )
+		{
+			/// \todo Can we reduce overhead by reading
```

Comment:
@donboie, do you have any insights here? My plan was to try to get my head around how the [C++ memory order](https://en.cppreference.com/w/cpp/atomic/memory_order) stuff applies here when I come to putting cancellation checks in the inner loops for the Algo. But if you have any thoughts in advance that would be great...

Reply:
> @donboie, do you have any insights here? 

Short answer is no. I did read a bunch of stuff to try to add some decent suggestions here but just came to the conclusion I didn't understand enough about memory ordering.

Reply:
Sounds like we're both in the same place then. Once we're dealing with the Algo side of things we'll at least have some practical experiments to inform us a bit...